core/thread/token.go

```diff
@@ -0,0 +1,189 @@
+package thread
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/gogo/status"
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/metautils"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	jwted25519 "github.com/textileio/go-threads/jwt"
+	"google.golang.org/grpc/codes"
+)
+
+// Identity represents an entity capable of signing a message
+// and returning the associated public key for verification.
+// In many cases, this will just be a private key, but callers
+// can use any setup that suits their needs.
+type Identity interface {
+	Sign([]byte) ([]byte, error)
+	GetPublic() PubKey
+}
+
+// Libp2pIdentity wraps crypto.PrivKey, overwriting GetPublic with thread.PubKey.
+type Libp2pIdentity struct {
+	crypto.PrivKey
+}
+
+// NewLibp2pIdentity returns a new Libp2pIdentity.
+func NewLibp2pIdentity(key crypto.PrivKey) Identity {
+	return &Libp2pIdentity{PrivKey: key}
+}
+
+func (p *Libp2pIdentity) GetPublic() PubKey {
+	return NewLibp2pPubKey(p.PrivKey.GetPublic())
+}
+
+// Pubkey can be anything that provides a verify method.
+type PubKey interface {
+	// String encodes the public key into a base64 string.
+	fmt.Stringer
+	// UnmarshalString decodes a public key from a base64 string.
+	UnmarshalString(string) error
+	// Verify that 'sig' is the signed hash of 'data'
+	Verify(data []byte, sig []byte) (bool, error)
+}
+
+// Libp2pPubKey wraps crypto.PubKey.
+type Libp2pPubKey struct {
+	crypto.PubKey
+}
+
+// NewLibp2pPubKey returns a new PubKey.
+func NewLibp2pPubKey(key crypto.PubKey) PubKey {
+	return &Libp2pPubKey{PubKey: key}
+}
+
+func (p *Libp2pPubKey) String() string {
+	bytes, err := crypto.MarshalPublicKey(p)
+	if err != nil {
+		panic(err)
+	}
+	return base64.StdEncoding.EncodeToString(bytes)
+}
+
+func (p *Libp2pPubKey) UnmarshalString(str string) error {
+	bytes, err := base64.StdEncoding.DecodeString(str)
+	if err != nil {
+		return err
+	}
+	p.PubKey, err = crypto.UnmarshalPublicKey(bytes)
+	return err
+}
+
+// Token is a concrete type for a JWT token string, which provides
+// a claim to an identity.
+type Token string
+
+// ErrTokenNotFound indicates the token was not found in the context.
+var ErrTokenNotFound = fmt.Errorf("thread token not found")
+
+// ErrInvalidToken indicates the token is invalid.
+var ErrInvalidToken = fmt.Errorf("invalid thread token")
+
+// NewToken issues a new JWT token from issuer for the given pubic key.
+func NewToken(issuer crypto.PrivKey, key PubKey) (tok Token, err error) {
+	var ok bool
+	issuer, ok = issuer.(*crypto.Ed25519PrivateKey)
+	if !ok {
+		log.Fatal("issuer must be an Ed25519PrivateKey")
+	}
+	claims := jwt.StandardClaims{
+		Subject:  key.String(),
+		Issuer:   NewLibp2pIdentity(issuer).GetPublic().String(),
+		IssuedAt: time.Now().Unix(),
```

Comment:
We could add an expiry here in the future to help limit the downside of a stolen token.